File: src/main/java/at/yawk/fiction/Chapter.java
package at.yawk.fiction;

import at.yawk.fiction.impl.fanfiction.FfnChapter;
import at.yawk.fiction.impl.fimfiction.FimChapter;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import lombok.Data;

/**
 * @author yawkat
 */
@Data
@JsonTypeInfo(use = JsonTypeInfo.Id.MINIMAL_CLASS)
@JsonSubTypes({
        @JsonSubTypes.Type(FfnChapter.class),
        @JsonSubTypes.Type(FimChapter.class),
})
public abstract class Chapter {
    String name;
    FormattedText text;
    /**
     * Whether this chapter is read according to account data of this user. {@code null} means unknown or not logged
     * in.
     */
    Boolean read;
}

File: src/main/kotlin/at/yawk/fiction/Chapter.kt
package at.yawk.fiction

import com.fasterxml.jackson.annotation.JsonTypeInfo

/**
 * @author yawkat
 */
data class Chapter(
        val name: String? = null,
        val text: FormattedText? = null,
        /**
         * Whether this chapter is read according to account data of this user. `null` means unknown or not logged
         * in.
         */
        val read: Boolean? = null,
        val providerData: ChapterProviderData
) : Mergeable<Chapter> {
    override fun merge(other: Chapter): Chapter {
        return Chapter(
                name = name ?: other.name,
                text = text ?: other.text,
                read = read ?: other.read,
                providerData = providerData.merge(other.providerData)
        )
    }
}

@JsonTypeInfo(use = JsonTypeInfo.Id.MINIMAL_CLASS)
interface ChapterProviderData : Mergeable<ChapterProviderData>, ProviderData